src/data-source/collection-browser-data-source.ts

```diff
@@ -0,0 +1,1189 @@
+      this.host.applyDefaultCollectionSort(this.collectionExtraInfo);
+
+      if (this.collectionExtraInfo) {
+        this.parentCollections = [].concat(
+          this.collectionExtraInfo.public_metadata?.collection ?? []
+        );
+      }
+    } else if (this.host.withinProfile) {
+      this.accountExtraInfo = success.response.accountExtraInfo;
+      this.pageElements = success.response.pageElements;
+    }
+
+    const { results, collectionTitles } = success.response;
+    if (results && results.length > 0) {
+      // Load any collection titles present on the response into the cache,
+      // or queue up preload fetches for them if none were present.
+      if (collectionTitles) {
+        for (const [id, title] of Object.entries(collectionTitles)) {
+          this.collectionTitles.set(id, title);
+        }
+      }
+
+      // Update the data source for each returned page
+      for (let i = 0; i < numPages; i += 1) {
+        const pageStartIndex = this.pageSize * i;
+        this.addTilesToDataSource(
+          pageNumber + i,
+          results.slice(pageStartIndex, pageStartIndex + this.pageSize)
+        );
+      }
+    }
+
+    // When we reach the end of the data, we can set the infinite scroller's
+    // item count to the real total number of results (rather than the
+    // temporary estimates based on pages rendered so far).
+    const resultCountDiscrepancy = numRows - results.length;
+    if (resultCountDiscrepancy > 0) {
+      this.endOfDataReached = true;
+      this.host.setTotalResultCount(this.totalResults);
+    }
+
+    for (let i = 0; i < numPages; i += 1) {
+      this.pageFetchesInProgress[pageFetchQueryKey]?.delete(pageNumber + i);
+    }
+
+    this.host.requestUpdate();
+  }
+
+  /**
+   * Update the datasource from the fetch response
+   *
+   * @param pageNumber
+   * @param results
+   */
+  private addTilesToDataSource(
+    pageNumber: number,
+    results: SearchResult[]
+  ): void {
+    // copy our existing datasource so when we set it below, it gets set
+    // instead of modifying the existing dataSource since object changes
+    // don't trigger a re-render
+    // const datasource = { ...this.dataSource };
+    const tiles: TileModel[] = [];
+    results?.forEach(result => {
+      if (!result.identifier) return;
+
+      let loginRequired = false;
+      let contentWarning = false;
+      // Check if item and item in "modifying" collection, setting above flags
+      if (
+        result.collection?.values.length &&
+        result.mediatype?.value !== 'collection'
+      ) {
+        for (const collection of result.collection?.values ?? []) {
+          if (collection === 'loggedin') {
+            loginRequired = true;
+            if (contentWarning) break;
+          }
+          if (collection === 'no-preview') {
+            contentWarning = true;
+            if (loginRequired) break;
+          }
+        }
+      }
+
+      tiles.push({
```

Comment:
q: can we do `tiles.push(new TileInfo(result))` and extract this data transform to another file? it'll be easier to test independently